File: 工具Tool/自定义系统控件/一些好用的SF插件/无限轮播/SFBannerScrollView/SFPageControl.m
//
//  SFPageControl.m
//  AdDemo
//
//  Created by lurich on 2021/6/11.
//

#import "SFPageControl.h"

/**
 *  Default number of pages for initialization
 */
static NSInteger const SFDefaultNumberOfPages = 0;

/**
 *  Default current page for initialization
 */
static NSInteger const SFDefaultCurrentPage = 0;

/**
 *  Default setting for hide for single page feature. For initialization
 */
static BOOL const SFDefaultHideForSinglePage = NO;

/**
 *  Default setting for shouldResizeFromCenter. For initialiation
 */
static BOOL const SFDefaultShouldResizeFromCenter = YES;

/**
 *  Default spacing between dots
 */
static NSInteger const SFDefaultSpacingBetweenDots = 8;

/**
 *  Default dot size
 */
static CGSize const SFDefaultDotSize = {8, 8};


static CGFloat const SFAnimateDuration = 1;

@implementation AbstractDotView


- (id)init {
    @throw [NSException exceptionWithName:NSInternalInconsistencyException
                                   reason:[NSString stringWithFormat:@"You must override %@ in %@", NSStringFromSelector(_cmd), self.class]
                                 userInfo:nil];
}


- (void)changeActivityState:(BOOL)active {
    @throw [NSException exceptionWithName:NSInternalInconsistencyException
                                   reason:[NSString stringWithFormat:@"You must override %@ in %@", NSStringFromSelector(_cmd), self.class]
                                 userInfo:nil];
}

@end

@implementation AnimatedDotView

- (instancetype)init {
    self = [super init];
    if (self) {
        [self initialization];
    }
    return self;
}

- (id)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self initialization];
    }
    return self;
}


- (id)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self initialization];
    }
    return self;
}

- (void)setDotColor:(UIColor *)dotColor {
    _dotColor = dotColor;
    self.layer.borderColor  = dotColor.CGColor;
}

- (void)initialization {
    _dotColor = [UIColor whiteColor];
    self.backgroundColor    = [UIColor clearColor];
    self.layer.cornerRadius = CGRectGetWidth(self.frame) / 2;
    self.layer.borderColor  = [UIColor whiteColor].CGColor;
    self.layer.borderWidth  = 2;
}


- (void)changeActivityState:(BOOL)active {
    if (active) {
        [self animateToActiveState];
    } else {
        [self animateToDeactiveState];
    }
}


- (void)animateToActiveState {
    [UIView animateWithDuration:SFAnimateDuration delay:0 usingSpringWithDamping:.5 initialSpringVelocity:-20 options:UIViewAnimationOptionCurveLinear animations:^{
        self.backgroundColor = self->_dotColor;
        self.transform = CGAffineTransformMakeScale(1.4, 1.4);
    } completion:nil];
}

- (void)animateToDeactiveState {
    [UIView animateWithDuration:SFAnimateDuration delay:0 usingSpringWithDamping:.5 initialSpringVelocity:0 options:UIViewAnimationOptionCurveLinear animations:^{
        self.backgroundColor = [UIColor clearColor];
        self.transform = CGAffineTransformIdentity;
    } completion:nil];
}

@end

@implementation DotView


- (instancetype)init {
    self = [super init];
    if (self) {
        [self initialization];
    }
    
    return self;
}

- (id)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self initialization];
    }
    return self;
}


- (id)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self initialization];
    }
    
    return self;
}

- (void)initialization {
    self.backgroundColor    = [UIColor clearColor];
    self.layer.cornerRadius = CGRectGetWidth(self.frame) / 2;
    self.layer.borderColor  = [UIColor whiteColor].CGColor;
    self.layer.borderWidth  = 2;
}


- (void)changeActivityState:(BOOL)active {
    if (active) {
        self.backgroundColor = [UIColor whiteColor];
    } else {
        self.backgroundColor = [UIColor clearColor];
    }
}

@end

@interface SFPageControl()

/**
 *  Array of dot views for reusability and touch events.
 */
@property (strong, nonatomic) NSMutableArray *dots;


@end

@implementation SFPageControl


#pragma mark - Lifecycle

- (id)init {
    self = [super init];
    if (self) {
        [self initialization];
    }
    
    return self;
}


- (id)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self initialization];
    }
    return self;
}


- (id)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self initialization];
    }
    
    return self;
}


/**
 *  Default setup when initiating control
 */
- (void)initialization {
    self.dotViewClass           = [AnimatedDotView class];
    self.spacingBetweenDots     = SFDefaultSpacingBetweenDots;
    self.numberOfPages          = SFDefaultNumberOfPages;
    self.currentPage            = SFDefaultCurrentPage;
    self.hidesForSinglePage     = SFDefaultHideForSinglePage;
    self.shouldResizeFromCenter = SFDefaultShouldResizeFromCenter;

    self.contentMode = UIViewContentModeScaleAspectFit;
}


#pragma mark - Touch event

- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {
    UITouch *touch = [touches anyObject];
    if (touch.view != self) {
        NSInteger index = [self.dots indexOfObject:touch.view];
        if ([self.delegate respondsToSelector:@selector(ddgPageControl:didSelectPageAtIndex:)]) {
            [self.delegate ddgPageControl:self didSelectPageAtIndex:index];
        }
    }
}

#pragma mark - Layout

/**
 *  Resizes and moves the receiver view so it just encloses its subviews.
 */
- (void)sizeToFit {
    [self updateFrame:YES];
}


- (CGSize)sizeForNumberOfPages:(NSInteger)pageCount {
    return CGSizeMake((self.dotSize.width + self.spacingBetweenDots) * pageCount - self.spacingBetweenDots , self.dotSize.height);
}


/**
 *  Will update dots display and frame. Reuse existing views or instantiate one if required. Update their position in case frame changed.
 */
- (void)updateDots {
    if (self.numberOfPages == 0) {
        return;
    }
    for (NSInteger i = 0; i < self.numberOfPages; i++) {
        
        UIView *dot;
        if (i < self.dots.count) {
            dot = [self.dots objectAtIndex:i];
        } else {
            dot = [self generateDotView];
        }
        [self updateDotFrame:dot atIndex:i];
    }
    [self changeActivity:YES atIndex:self.currentPage];
    [self hideForSinglePage];
}

/**
 *  Update frame control to fit current number of pages. It will apply required size if authorize and required.
 *
 *  @param overrideExistingFrame BOOL to allow frame to be overriden. Meaning the required size will be apply no mattter what.
 */
- (void)updateFrame:(BOOL)overrideExistingFrame {
    CGPoint center = self.center;
    CGSize requiredSize = [self sizeForNumberOfPages:self.numberOfPages];
    
    // We apply requiredSize only if authorize to and necessary
    if (overrideExistingFrame || ((CGRectGetWidth(self.frame) < requiredSize.width || CGRectGetHeight(self.frame) < requiredSize.height) && !overrideExistingFrame)) {
        self.frame = CGRectMake(CGRectGetMinX(self.frame), CGRectGetMinY(self.frame), requiredSize.width, requiredSize.height);
        if (self.shouldResizeFromCenter) {
            self.center = center;
        }
    }
    [self resetDotViews];
}

/**
 *  Update the frame of a specific dot at a specific index
 *
 *  @param dot   Dot view
 *  @param index Page index of dot
 */
- (void)updateDotFrame:(UIView *)dot atIndex:(NSInteger)index
{
    // Dots are always centered within view
    CGFloat x = (self.dotSize.width + self.spacingBetweenDots) * index + ( (CGRectGetWidth(self.frame) - [self sizeForNumberOfPages:self.numberOfPages].width) / 2);
    CGFloat y = (CGRectGetHeight(self.frame) - self.dotSize.height) / 2;
    dot.frame = CGRectMake(x, y, self.dotSize.width, self.dotSize.height);
}

#pragma mark - Utils

/**
 *  Generate a dot view and add it to the collection
 *
 *  @return The UIView object representing a dot
 */
- (UIView *)generateDotView {
    UIView *dotView;
    
    if (self.dotViewClass) {
        dotView = [[self.dotViewClass alloc] initWithFrame:CGRectMake(0, 0, self.dotSize.width, self.dotSize.height)];
        if ([dotView isKindOfClass:[AnimatedDotView class]] && self.dotColor) {
            ((AnimatedDotView *)dotView).dotColor = self.dotColor;
        }
    } else {
        dotView = [[UIImageView alloc] initWithImage:self.dotImage];
        dotView.frame = CGRectMake(0, 0, self.dotSize.width, self.dotSize.height);
        dotView.contentMode = UIViewContentModeScaleAspectFit;
    }
    if (dotView) {
        [self addSubview:dotView];
        [self.dots addObject:dotView];
    }
    dotView.userInteractionEnabled = YES;
    return dotView;
}

/**
 *  Change activity state of a dot view. Current/not currrent.
 *
 *  @param active Active state to apply
 *  @param index  Index of dot for state update
 */
- (void)changeActivity:(BOOL)active atIndex:(NSInteger)index {
    if (self.dotViewClass) {
        AbstractDotView *abstractDotView = (AbstractDotView *)[self.dots objectAtIndex:index];
        if ([abstractDotView respondsToSelector:@selector(changeActivityState:)]) {
            [abstractDotView changeActivityState:active];
        } else {
            NSLog(@"Custom view : %@ must implement an 'changeActivityState' method or you can subclass %@ to help you.", self.dotViewClass, [AbstractDotView class]);
        }
    } else if (self.dotImage && self.currentDotImage) {
        UIImageView *dotView = (UIImageView *)[self.dots objectAtIndex:index];
        dotView.image = (active) ? self.currentDotImage : self.dotImage;
    }
}

- (void)resetDotViews {
    for (UIView *dotView in self.dots) {
        [dotView removeFromSuperview];
    }
    
    [self.dots removeAllObjects];
    [self updateDots];
}


- (void)hideForSinglePage {
    if (self.dots.count == 1 && self.hidesForSinglePage) {
        self.hidden = YES;
    } else {
        self.hidden = NO;
    }
}

#pragma mark - Setters

- (void)setNumberOfPages:(NSInteger)numberOfPages {
    _numberOfPages = numberOfPages;
    
    // Update dot position to fit new number of pages
    [self resetDotViews];
}

- (void)setSpacingBetweenDots:(NSInteger)spacingBetweenDots {
    _spacingBetweenDots = spacingBetweenDots;
    [self resetDotViews];
}

- (void)setCurrentPage:(NSInteger)currentPage {
    // If no pages, no current page to treat.
    if (self.numberOfPages == 0 || currentPage == _currentPage) {
        _currentPage = currentPage;
        return;
    }
    // Pre set
    [self changeActivity:NO atIndex:_currentPage];
    _currentPage = currentPage;
    // Post set
    [self changeActivity:YES atIndex:_currentPage];
}

- (void)setDotImage:(UIImage *)dotImage {
    _dotImage = dotImage;
    [self resetDotViews];
    self.dotViewClass = nil;
}

- (void)setCurrentDotImage:(UIImage *)currentDotimage {
    _currentDotImage = currentDotimage;
    [self resetDotViews];
    self.dotViewClass = nil;
}

- (void)setDotViewClass:(Class)dotViewClass {
    _dotViewClass = dotViewClass;
    self.dotSize = CGSizeZero;
    [self resetDotViews];
}

#pragma mark - Getters

- (NSMutableArray *)dots {
    if (!_dots) {
        _dots = [[NSMutableArray alloc] init];
    }
    return _dots;
}


- (CGSize)dotSize{
    // Dot size logic depending on the source of the dot view
    if (self.dotImage && CGSizeEqualToSize(_dotSize, CGSizeZero)) {
        _dotSize = self.dotImage.size;
    } else if (self.dotViewClass && CGSizeEqualToSize(_dotSize, CGSizeZero)) {
        _dotSize = SFDefaultDotSize;
        return _dotSize;
    }
    return _dotSize;
}

@end

File: 第三方/Swift/SwiftRunLoop/SwiftRunLoop/SFRunLoop/SFPageControl.swift
//
//  SFPageControl.swift
//  SwiftRunLoop
//
//  Created by lurich on 2022/1/21.
//

import Foundation
import UIKit

class SFAnimatedDotView: UIView {
    var dotColor: UIColor {
        didSet {
            layer.borderColor = dotColor.cgColor
        }
    }
    
    override init(frame: CGRect) {
        dotColor = UIColor.white
        super.init(frame: frame)
        backgroundColor = UIColor.clear
        layer.masksToBounds = true
        layer.cornerRadius = frame.size.width / 2.0
        layer.borderColor = UIColor.white.cgColor
        layer.borderWidth = 2.0
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func changeActivityState(_ active: Bool) {
        if active {
            animateToActiveState()
        } else {
            animateToDeactiveState()
        }
    }
    
    func animateToActiveState() {
        UIView.animate(withDuration: 1.0, delay: 0, usingSpringWithDamping: 0.5, initialSpringVelocity: 0, options: .curveLinear) { [self] in
            self.backgroundColor = self.dotColor
            self.transform = CGAffineTransform(scaleX: 1.3, y: 1.3)
        }
    }
    
    func animateToDeactiveState() {
        UIView.animate(withDuration: 1.0, delay: 0, usingSpringWithDamping: 0.5, initialSpringVelocity: 0, options: .curveLinear) { [self] in
            self.backgroundColor = UIColor.clear
            self.transform = .identity
        }
    }
}

class SFPageControl: UIControl {
    var dotView: SFAnimatedDotView?
    var dotImage: UIImage? {
        didSet {
            if self.dotImage != nil {
                self.dotSize = self.dotImage!.size
            }
            self.dotView = nil
        }
    }
    var currentDotImage: UIImage? {
        didSet {
            self.dotView = nil
        }
    }
    var dotSize: CGSize
    var dotColor: UIColor? 
    var spacingBetweenDots: Double
    var numberOfPages: Int
    var currentPage: Int {
        willSet {
            if self.numberOfPages == 0 {
                return
            }
            change(activity: false, atIndex: self.currentPage)
        }
        didSet {
            if self.numberOfPages == 0 {
                return
            }
            change(activity: true, atIndex: self.currentPage)
        }
    }
    var dots: Array<UIView>
    
    override init(frame: CGRect) {
        dotView = SFAnimatedDotView()
        dotSize = CGSize(width: 8, height: 8)
        dotColor = UIColor.white
        spacingBetweenDots = 8.0
        numberOfPages = 0
        currentPage = 0
        dots = [UIView]()
        super.init(frame: frame)
        contentMode = .scaleAspectFit
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func animatedPageSize() -> CGSize {
        CGSize(width: (self.dotSize.width + self.spacingBetweenDots) * Double(self.numberOfPages) - self.spacingBetweenDots, height: self.dotSize.height)
    }
    
    func updateDots() {
        if self.numberOfPages == 0 {
            return
        }
        for i in 0..<self.numberOfPages {
            var dot: UIView
            if i < self.dots.count {
                dot = self.dots[i]
            } else {
                dot = generateDotView()
            }
            updateDot(frame: dot, atIndex: i)
        }
        change(activity: true, atIndex: self.currentPage)
    }
    
    func updateFrame() {
        let requiredSize = animatedPageSize()
        self.frame = CGRect(origin: self.frame.origin, size: requiredSize)
        resetDotViews()
    }
    
    func updateDot(frame dot: UIView, atIndex index: Int) {
        let x = (self.dotSize.width + self.spacingBetweenDots) * Double(index) + (self.frame.size.width - animatedPageSize().width) / 2.0
        let y = (self.frame.size.height - self.dotSize.height) / 2.0
        dot.frame = CGRect(origin: CGPoint(x: x, y: y), size: self.dotSize)
    }
    
    func generateDotView() -> UIView {
        var dotView: UIView
        if self.dotView != nil {
            dotView = SFAnimatedDotView(frame: CGRect(origin: .zero, size: self.dotSize))
            if self.dotColor != nil {
                (dotView as! SFAnimatedDotView).dotColor = self.dotColor!
            }
        } else {
            dotView = UIImageView(frame: CGRect(origin: .zero, size: self.dotSize))
            (dotView as! UIImageView).image = self.dotImage
            dotView.contentMode = .scaleAspectFit
        }
        self.addSubview(dotView)
        self.dots.append(dotView)
        dotView.isUserInteractionEnabled = true
        return dotView
    }
    
    func change(activity active: Bool, atIndex index: Int) {
        if self.dotView != nil {
            let abstractDotView = self.dots[index] as! SFAnimatedDotView
            abstractDotView.changeActivityState(active)
        } else if (self.dotImage != nil && self.currentDotImage != nil) {
            let dotView = self.dots[index] as! UIImageView
            dotView.image = active ? self.currentDotImage : self.dotImage
        }
    }
    
    func resetDotViews() {
        for dotView in self.dots {
            dotView.removeFromSuperview()
        }
        self.dots.removeAll()
        updateDots()
    }
}
